File: Tree/226. Invert Binary Tree.cpp
/*
226. Invert Binary Tree

Invert a binary tree.

Example:

Input:

     4
   /   \
  2     7
 / \   / \
1   3 6   9
Output:

     4
   /   \
  7     2
 / \   / \
9   6 3   1


*/


class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if(root) {
            swap(root->left, root->right);
            invertTree(root->left);
            invertTree(root->right);
        }
        return root;
    }
};

class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if(root) {
            TreeNode* tmp = root->left;
            root->left = invertTree(root->right);
            root->right = invertTree(tmp);
            
        }
        return root;
    }
};

//DFS
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if(!root) return root;
        stack<TreeNode*>stk;
        stk.push(root);
        while(!stk.empty()){
            TreeNode* cur = stk.top();
            stk.pop();
            swap(cur->left, cur->right);
            if(cur->left) stk.push(cur->left);
            if(cur->right) stk.push(cur->right);
        }
        return root;
    }
};

//BFS,// Space: O(w), w is the max number of nodes of the levels.
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if(!root) return root;
        queue<TreeNode*>stk;
        stk.push(root);
        while(!stk.empty()){
            TreeNode* cur = stk.front();
            stk.pop();
            swap(cur->left, cur->right);
            if(cur->left) stk.push(cur->left);
            if(cur->right) stk.push(cur->right);
        }
        return root;
    }
};

File: Tree/226. Invert Binary Tree.py
"""
226. Invert Binary Tree

Invert a binary tree.

Example:

Input:

     4
   /   \
  2     7
 / \   / \
1   3 6   9
Output:

     4
   /   \
  7     2
 / \   / \
9   6 3   1

"""

class Solution:
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        def revert(node):
            if not node: return 
            node.left, node.right = node.right, node.left
            revert(node.left)
            revert(node.right)
        revert(root)
        return root

class Solution:
    def invertTree(self, root):
        if not root: return None
        root.left,root.right = self.invertTree(root.right), self.invertTree(root.left)
        return root



class Solution: 
    def invertTree(self, root):
        if root:
            invert = self.invertTree
            root.left, root.right = invert(root.right), invert(root.left)
            return root

class Solution:
    def invertTree(self, root):
        queue = collections.deque([(root)])
        while queue:
            node = queue.popleft()
            if node:
                node.left, node.right = node.right, node.left
                queue.append(node.left)
                queue.append(node.right)
        return root

class Solution:
    def invertTree(self, root):
        stack = [root]
        while stack:
            node = stack.pop()
            if node:
                node.left, node.right = node.right, node.left
                stack += node.left, node.right
        return root